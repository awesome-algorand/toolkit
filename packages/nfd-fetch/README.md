# @awesome-algorand/nfd-fetch

Experimental package for fetching data from NF Domains api.

## Building the client

Generate the TypeScript client using HeyApi

```bash
npm run generate
```

Build the library using

```bash
npm run build
```

## Usage

Install the library

```bash
npm install @awesome-algorand/nfd-fetch  @hey-api/client-fetch --save
```

```typescript
import {nfdBrowse, type NfdRecordCollection} from "@awesome-algorand/nfd-fetch";

await nfdBrowse({limit: 10, offset: 0, sort: "asc", order: "id"}).then((data: NfdRecordCollection) => {
    console.log(data);
}, {baseUrl: "https://api.nf.domains"});
```
