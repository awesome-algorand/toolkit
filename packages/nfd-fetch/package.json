{
  "name": "@awesome-algorand/nfd-fetch",
  "description": "Generated fetch library for NF Domains",
  "version": "0.0.1-experimental.1",
  "main": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "files": [
    "lib",
    "src"
  ],
  "exports": {
    ".": {
      "types": "./lib/index.d.ts",
      "default": "./lib/index.js"
    }
  },
  "scripts": {
    "build": "tsc",
    "generate": "npx @hey-api/openapi-ts -c @hey-api/client-fetch -i https://api.nf.domains/info/openapi3.yaml -o ./src/",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "Awesome Algorand",
  "license": "Unlicense",
  "peerDependencies": {
    "@hey-api/client-fetch": "^0.7.2"
  },
  "devDependencies": {
    "@hey-api/client-fetch": "^0.7.2",
    "typescript": "^5.7.3"
  }
}
