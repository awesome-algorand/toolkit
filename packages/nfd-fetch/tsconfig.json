{
  "compilerOptions": {
    "strict": true,
    "ignoreDeprecations": "5.0",
    "allowJs": true,
    "skipLibCheck": true,
    "isolatedModules": true,
    "moduleResolution": "node",
    "module": "esnext",
    "target": "esnext",
    "sourceMap": true,
    "declaration": true,
    "outDir": "./lib"
  },
  "include": [
    "src/**/*.ts"
  ],
  "typedocOptions": {
    "entryPoints": [
      "src/index.ts"
    ],
    "out": "./docs"
  }
}
