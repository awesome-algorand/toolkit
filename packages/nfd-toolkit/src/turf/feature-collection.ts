import {NfdRecordCollection} from "@awesome-algorand/nfd-fetch";
import type { FeatureCollection, Feature, Point } from "geojson";
import {NfdGeoProperties, toProperties} from "./properties.js";
import {toPosition} from "./position.js";

/**
 * Converts an NfdRecordCollection and a cache of existing features into a GeoJSON FeatureCollection.
 *
 * @param {NfdRecordCollection} nfd - The collection of NFD records to be transformed into a FeatureCollection.
 * @param {Feature<Point, NfdGeoProperties>[]} cache - An array of cached GeoJSON features used to optimize processing and avoid redundant data retrieval.
 * @return {Promise<FeatureCollection<Point, NfdGeoProperties>>} A promise that resolves to a GeoJSON FeatureCollection containing features derived from the input NFD records.
 */
export async function toFeatureCollection(nfd: NfdRecordCollection, cache: Feature<Point, NfdGeoProperties>[] = []): Promise<FeatureCollection<Point, NfdGeoProperties>>{
    return {
        type: "FeatureCollection",
        features: await Promise.all(
            nfd.map(async (record) => {
                const results = cache.find((point) => point.properties && point.properties.name === record.name);
                let position = typeof results !== 'undefined' ? results.geometry.coordinates : await toPosition(record?.properties);
                return {
                    type: "Feature",
                    geometry: {
                        type: "Point",
                        coordinates: position,
                    },
                    properties: await toProperties(record),
                };
            }),
        ),
    };
}
