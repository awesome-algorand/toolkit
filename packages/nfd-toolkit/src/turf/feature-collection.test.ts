import {expect, test, describe} from 'vitest'
import {toFeatureCollection} from "./feature-collection";

import fullFixtures from '../../test/fixtures/full.fixtures.json'
import fullGolden from './feature-collection.test.full.golden.json'
// TODO: other fixtures
// import briefFixtures from '../fixtures/brief.fixtures.json'
// import tinyFixtures from '../fixtures/tiny.fixtures.json'
import {NfdRecordCollection} from "@awesome-algorand/nfd-fetch";
import {createMockFetch} from "../../test/mock";


describe("Feature Collection", () => {
    if(typeof process.env.INTEGRATION === 'undefined'){
        global.fetch = createMockFetch()
    }
    test("toFeatureCollection", async () => {
        expect(await toFeatureCollection([])).toEqual({
            "features": [],
            "type": "FeatureCollection",
        })
        expect(await toFeatureCollection(fullFixtures as unknown as NfdRecordCollection)).toEqual(fullGolden)
    }, 10000)


})

