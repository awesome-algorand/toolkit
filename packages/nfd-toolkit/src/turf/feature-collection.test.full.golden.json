{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "geometry": {
        "type": "Point",
        "coordinates": [
          -76.9382069,
          39.5162401
        ]
      },
      "properties": {
        "name": "phear.algo",
        "marker-color": "#ff602e",
        "timeChanged": "2024-11-19T00:38:48Z",
        "avatar": {
          "verified": false,
          "url": "https://images.nf.domains/avatar/a1481ef6-ee4a-4971-99e9-32a8bdef001f"
        },
        "banner": {
          "verified": false,
          "url": "https://images.nf.domains/banner/878fcfdc-61b0-4caa-9e80-350080838969"
        }
      }
    },
    {
      "type": "Feature",
      "geometry": {
        "type": "Point",
        "coordinates": [
          -105.270545,
          40.0149856
        ]
      },
      "properties": {
        "name": "gabe.algo",
        "marker-color": "#ff602e",
        "timeChanged": "2025-01-16T05:32:32Z",
        "avatar": {
          "verified": true,
          "url": "https://ipfs.algonode.xyz/ipfs/QmfEZK3nVHJGDQqjG8wEjrNMghQFdXygW8u7yyu2eVgtRu"
        },
        "banner": {
          "verified": true,
          "url": "https://ipfs.algonode.xyz/ipfs/bafybeiag7yek2wu3ahuzewygi2gaxj22mn3zcyaccxlgplg2lfsus6222q"
        }
      }
    },
    {
      "type": "Feature",
      "geometry": {
        "type": "Point",
        "coordinates": [
          151.2082848,
          -33.8698439
        ]
      },
      "properties": {
        "name": "adri.algo",
        "marker-color": "#ff602e",
        "timeChanged": "2024-12-19T09:31:34Z",
        "avatar": {
          "verified": true,
          "url": "https://ipfs.algonode.xyz/ipfs/QmYzsfJGgkp8baa7TjTFURbEkFD9L6qGwMhWQXUAVQVkfx"
        },
        "banner": {
          "verified": false,
          "url": "https://images.nf.domains/banner/4c04dee3-a5d3-451e-a02f-54690a45f923"
        }
      }
    },
    {
      "type": "Feature",
      "geometry": {
        "type": "Point",
        "coordinates": [
          19.134422,
          52.215933
        ]
      },
      "properties": {
        "name": "urtho.algo",
        "marker-color": "#ff602e",
        "timeChanged": "2024-11-18T19:49:34Z",
        "avatar": {
          "verified": false,
          "url": "https://images.nf.domains/avatar/15cd547f-10ac-46bb-bc63-67ec89b6e18e"
        },
        "banner": {
          "verified": false,
          "url": "https://images.nf.domains/banner/eb82cab3-7069-40f1-aaff-6a79b8833743"
        }
      }
    },
    {
      "type": "Feature",
      "geometry": {
        "type": "Point",
        "coordinates": [
          -75.0458515,
          -6.8699697
        ]
      },
      "properties": {
        "name": "evertdiaz.directory.algo",
        "marker-color": "#ff602e",
        "timeChanged": "2024-12-17T01:06:51Z",
        "avatar": {
          "verified": false,
          "url": "https://images.nf.domains/avatar/ca8cd126-79c0-4ad4-9ec3-e5bdf84244b3"
        }
      }
    },
    {
      "type": "Feature",
      "geometry": {
        "type": "Point",
        "coordinates": [
          -40,
          25
        ]
      },
      "properties": {
        "name": "evertdiaz.algo",
        "marker-color": "#ff602e",
        "timeChanged": "2023-07-05T22:19:14Z"
      }
    }
  ]
}
